src/gt4py/next/iterator/type_inference.py

```diff
@@ -477,6 +477,39 @@ def visit_FunCall(
                 )
                 constraints.add((tup, val))
                 return Val(kind=kind, dtype=elem, size=size)
+            if node.fun.id == "cast_":
+                if len(node.args) != 2:
+                    raise TypeError("cast_ requires exactly two arguments.")
+                val_arg_type = self.visit(node.args[0], constraints=constraints, symtypes=symtypes)
+                type_arg = node.args[1]
+                if not isinstance(type_arg, ir.SymRef) or type_arg.id not in ir.TYPEBUILTINS:
+                    raise TypeError("The second argument to `cast_` must be a type literal.")
+
+                kind = TypeVar.fresh()
+                size = TypeVar.fresh()
+                current_loc = TypeVar.fresh()
+                defined_loc = TypeVar.fresh()
+
+                constraints.add(
+                    (
+                        val_arg_type,
+                        Val(
+                            kind=kind,
```

Comment:
```suggestion
                            kind=Value(),
```
since cast only works on values?

Reply:
Good point.